additional edge cases tests for `path.rs` 🧪

GrantBirki opened the pull request:
This pull request adds a few new edge case tests to the `std::path` module. The new tests cover scenarios such as paths with only separators, non-ASCII and Unicode characters, embedded new lines, etc. Each new test is documented with some helpful in-line comments as well.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

GrantBirki committed:
additional edge cases tests for `path.rs`

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15078447747/job/42390863222) [(plain)](https://github.com/rust-lang/rust/commit/49658e6a3c528dec342d82920f4606dab4d90ae8/checks/42390863222/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.790 Building wheels for collected packages: reuse
#12 2.792   Building wheel for reuse (pyproject.toml): started
#12 3.002   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.003   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#12 3.003   Stored in directory: /tmp/pip-ephem-wheel-cache-muxf0kwl/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.005 Successfully built reuse
#12 3.006 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.399 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.399 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 3.923 Collecting virtualenv
#12 3.958   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.031      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 86.4 MB/s eta 0:00:00
#12 4.090 Collecting filelock<4,>=3.12.2
#12 4.093   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.127 Collecting platformdirs<5,>=3.9.1
#12 4.130   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.147 Collecting distlib<1,>=0.3.7
#12 4.159   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.166      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 100.8 MB/s eta 0:00:00
#12 4.247 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.436 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.436 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.5s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      161728 kB
DirectMap2M:     8226816 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 34.674
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6014 files
Diff in /checkout/library/std/tests/path.rs:10:
 use std::ffi::OsStr;
 use std::hash::{DefaultHasher, Hash, Hasher};
 use std::mem::MaybeUninit;
-use std::path::*;
-use std::ptr;
-use std::rc::Rc;
-use std::sync::Arc;
```

</details>

GrantBirki commented:
It appears that `mingw-check-tidy` failed, but I am not sure how to correct it. Perhaps its a linter of some sort?

ChaiTRex commented:
> It appears that `mingw-check-tidy` failed, but I am not sure how to correct it. Perhaps its a linter of some sort?

`mingw-check` also failed. Click on `CI / PR - mingw-check (pull_request)` and see the compiler errors toward the bottom.

ChaiTRex commented:
It looks as if `OsStrExt` and `OsStringExt` have different methods on Windows than other platforms.

workingjubilee commented:
Yes, please fix the tests to only use functions from OsStr and OsString OR cfg the tests.

workingjubilee commented on library/std/tests/path.rs:
```diff
@@ -1976,3 +1983,102 @@ fn clone_to_uninit() {
+    let bytes = b"foo\0bar";
+    let os_str = OsStr::from_bytes(bytes);
+    let path = Path::new(os_str);
+    assert!(path.as_os_str().as_bytes().contains(&0));
```
This relies on [OsStrExt](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html#tymethod.as_bytes)

GrantBirki replied:
It might be a little much so I just removed it for now. 🙇

GrantBirki committed:
remove `test_embedded_null_byte()` test for now

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15082202433/job/42400299513) [(plain)](https://github.com/rust-lang/rust/commit/604f0e27436b4940399873d0c23e5b609a52e831/checks/42400299513/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 4.286      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 25.8 MB/s eta 0:00:00
#12 4.330 Collecting distlib<1,>=0.3.7
#12 4.338   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.349      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 49.0 MB/s eta 0:00:00
#12 4.388 Collecting platformdirs<5,>=3.9.1
#12 4.395   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.436 Collecting filelock<4,>=3.12.2
#12 4.444   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.527 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.721 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.721 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      141248 kB
DirectMap2M:     9295872 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 35.411
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6014 files
Diff in /checkout/library/std/tests/path.rs:1998:
     assert_eq!(path.file_name(), Some(OsStr::new("file.txt")));
 }
 
-
 // Test: Reserved device names (Windows)
 // This test ensures that reserved device names like "CON", "PRN", etc., are handled as normal paths on non-Windows platforms,
 // and as special cases on Windows (if applicable).
Build completed unsuccessfully in 0:01:15
  local time: Sat May 17 05:51:16 UTC 2025
  network time: Sat, 17 May 2025 05:51:16 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

ChrisDenton commented on library/std/tests/path.rs:
```diff
@@ -1976,3 +1976,89 @@ fn clone_to_uninit() {
+    assert_eq!(path.file_name(), Some(OsStr::new(".ext")));
+}
+
+// Test: Long components
+// This test checks that Path can handle very long path components without truncation or error.
+// It ensures that the length of the component is preserved.
```
I'm not sure why this would be useful for us specifically? We don't usually check that slice wrappers work on the whole slice.

GrantBirki replied:
I'll remove this test and a few others that we think aren't the most useful perhaps

GrantBirki committed:
revert forward slash to backslash

GrantBirki commented on library/std/tests/path.rs:
```diff
@@ -1976,3 +1976,89 @@ fn clone_to_uninit() {
+    assert_eq!(path.to_str(), Some("bar..."));
+}
+
+// Test: Only extension (e.g., ".gitignore")
+// This test verifies that files with only an extension and no base name are handled correctly.
+// It checks that the extension is recognized and the file stem is None or empty as appropriate.
```
Actually I think you are right and this is tested already here: https://github.com/rust-lang/rust/blob/2c12b4a690c8201a3c79f404eec38614684c412c/library/std/tests/path.rs#L1127-L1130

GrantBirki committed:
remove extra tests that really might not be all that useful

GrantBirki commented:
@ChrisDenton I went ahead and removed a few of the tests that I added which feel rather superfluous now after reading your comments. This slims down the PR to really just be three new unit tests.

jhpratt commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 4358a1c05e8a8878edf22c8cb1f3eae07e09efc0 has been approved by `jhpratt`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jhpratt 4358a1c05e8a8878edf22c8cb1f3eae07e09efc0 -->
<!-- homu: {"type":"Approved","sha":"4358a1c05e8a8878edf22c8cb1f3eae07e09efc0","approver":"jhpratt","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
